Allow multiple raster selection from GDAL source select widget

elpaso opened the pull request:
Since I'm using QgsFileWidget I also added
multiple files support to that widget and
tests for this new behavior.

elpaso committed:
Allow multiple raster selection from GDAL source select widget

Since I'm using QgsFileWidget I also added
multiple files support to that widget and
tests for this new behavior.

nyalldawson commented on src/gui/qgsfilewidget.cpp:
```diff
@@ -231,25 +231,40 @@ void QgsFileWidget::openFileDialog()
     title = !mDialogTitle.isEmpty() ? mDialogTitle : tr( "Select a file" );
     fileName = QFileDialog::getOpenFileName( this, title, QFileInfo( oldPath ).absoluteFilePath(), mFilter );
   }
+  else if ( mStorageMode == GetMultipleFiles )
```
Could you swap all these ifs to switches? (With no default cases). It makes maintenance easier.

nyalldawson commented on src/gui/qgsfilewidget.cpp:
```diff
@@ -231,25 +231,40 @@ void QgsFileWidget::openFileDialog()
+  {
+    for ( int i = 0; i < fileNames.length(); i++ )
+    {
+      fileNames.replace( i, QDir::cleanPath( QFileInfo( fileNames.at( i ) ).absoluteFilePath() ) ) ;
```
Was the omission of toNativeSeparators deliberate here? It makes things much nicer for windows users.

elpaso replied:
My mistake, thanks for pointing this out.

nyalldawson commented on src/gui/qgsfilewidget.cpp:
```diff
@@ -359,16 +388,30 @@ void QgsFileDropEdit::setFilters( const QString &filters )
-           ( mAcceptableExtensions.isEmpty() || mAcceptableExtensions.contains( file.suffix(), Qt::CaseInsensitive ) ) )
-         || ( mStorageMode == QgsFileWidget::GetDirectory && file.isDir() ) )
-      path = file.filePath();
+    for ( const auto url : event->mimeData()->urls() )
```
Pretty sure this causes a detach and full copy. I think you need to wrap the event->mimeData()->urls() in a qAsConst(...)

nyalldawson replied:
Just a follow up - we can't use qAsConst since it's qt 5.7 and above only. I think we need to use std::as_const, and add a reference implementation of this for C++ <17 compilers (see http://en.cppreference.com/w/cpp/utility/as_const )

elpaso replied:
To be honest I knew it caused a full copy, but in this case I don't think we have any noticeable overhead. BTW, a `Q_FOREACH` would be acceptable?

nyalldawson replied:
Yeah, go with Q_FOREACH for now. I'm working on the as_const stuff, so let's see how that goes...

nyalldawson reviewed (commented):
This is great - thanks for investing in these reusable widgets!

nyalldawson commented:
One other thought - we should set a tooltip when multiple files are selected to show the files in a bulleted list. Otherwise it's tricky to see all the selected files if they are only shown in the single line lineedit.

elpaso committed:
Small code style changes

elpaso committed:
Do not quote when single raster file is selected

elpaso committed:
More robust multiple paths split and utility static method

elpaso commented:
@nyalldawson I think I've addressed all issues, added the tooltip and polished the code.
I've added an utility static method to split the multiple files, with test case.

elpaso commented:
@nyalldawson thanks for the review!

nirvn commented:
@elpaso , great, I was days away from filing an issue on this 😉 

Could the OGR (i.e. vector) source select also rely on QgsFileWidget? It'd harmonize the UI (i.e. button '...' vs 'browse' label, etc.), and it's just be good practice to have those two share the same file widget.
